File: Java/expressao/src/br/com/calculo/enums/Operacao.java
package br.com.calculo.enums;

import br.com.calculo.Expressao;
import br.com.calculo.ExpressaoBase;
import br.com.calculo.Soma;
import br.com.calculo.Subtracao;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

public enum Operacao {
    Soma("+") {
        @Override
        public Expressao calculo(Expressao numero1, Expressao numero2) {
            return new Soma(numero1, numero2);
        }

        @Override
        public Integer prioridade() {
            return 1;
        }
    }, Subtracao("-") {
        @Override
        public Expressao calculo(Expressao numero1, Expressao numero2) {
            return new Subtracao(numero1, numero2);
        }

        @Override
        public Integer prioridade() {
            return 1;
        }
    }, Divisao("/") {
        @Override
        public Expressao calculo(Expressao numero1, Expressao numero2) {
            return new ExpressaoBase(numero1, numero2) {

                @Override
                public Double calculo() {

                    return new BigDecimal(numero1.calculo().toString()).divide(new BigDecimal(numero2.calculo().toString()), BigDecimal.ROUND_HALF_DOWN).doubleValue();
                }
            };
        }

        @Override
        public Integer prioridade() {
            return 2;
        }
    }, Multiplicacao("*") {
        @Override
        public Expressao calculo(Expressao numero1, Expressao numero2) {
            return new ExpressaoBase(numero1, numero2) {
                @Override
                public Double calculo() {
                    return numero1.calculo() * numero2.calculo();
                }
            };
        }

        @Override
        public Integer prioridade() {
            return 2;
        }
    };

    private static Map<String, Operacao> map = new HashMap<>();

    static {
        for (Operacao op : Operacao.values()) {
            map.put(op.operador, op);
        }
    }


    public final String operador;

    private Operacao(String operador) {
        this.operador = operador;
    }

    public abstract Expressao calculo(Expressao numero1, Expressao numero2);

    public abstract Integer prioridade();

    public static boolean isOperador(String op) {
        return Operacao.map.containsKey(op);
    }

    public static Operacao getOperador(String op) {
        return Operacao.map.get(op);
    }
}

File: Kotlin/calculadoraKotlin/src/calculadora/Operacao.kt
package calculadora

import java.text.NumberFormat

class Soma(val expressao1: Expressao? = null, val  expressao2: Expressao? = null) : Expressao {

    override fun calcula(): Double = expressao1?.calcula()!!.plus(expressao2!!.calcula())

}

enum class Operacao(val operador: String) : Expressao {
    Soma("+") {
        override fun calcula(): Double {
            return Soma().calcula() //To change body of created functions use File | Settings | File Templates.
        }


    }
    /*Subtracao("-") {
        override fun calculo(): ExpressaoBase = object : ExpressaoBase() {
            override fun calcula(): Double = expressao1?.calcula()!!.minus(expressao2!!.calcula())
        }
    },
    Multiplicacao(operador = "X") {
        override fun calculo(): ExpressaoBase = object : ExpressaoBase() {
            override fun calcula(): Double = expressao1?.calcula()!!.times(expressao2!!.calcula())
        }
    },
    Divisao("/") {
        override fun calculo(): ExpressaoBase = object : ExpressaoBase() {
            override fun calcula(): Double = expressao1?.calcula()!!.div(expressao2!!.calcula())

        }
    }, Porcentagem(operador = "%"){
        override fun calculo(): ExpressaoBase = object : ExpressaoBase() {
            override fun calcula(): Double {
                return expressao1?.calcula()!!.times(expressao2!!.calcula().div(100))
            }
        }
    }*/;


    //abstract fun calculo(): Expressao

    companion object {
        fun from(findValue: String): Operacao? {

            val single = Operacao.values().singleOrNull({ ope -> ope.operador == findValue })
            return single
        }
    }
}